### PR TITLE
Add support for text shadows.

### PR DESCRIPTION
**Pull request by GaryQian**
This changes BoxShadow to inherit from Shadow (a new class without the spreadRadius)

TextStyle holds a `List<Shadow>`.

Fixes/Adds features for https://github.com/flutter/flutter/issues/3402

### CODE REVIEW

**Comment by GaryQian** [comment]
See https://github.com/flutter/engine/pull/6385 for Engine-side changes.

**Comment by GaryQian** [comment]
This is still missing golden tests, but I wanted to do that in a separate patch because to keep the merging of this one smooth.

File: packages/flutter/lib/src/painting/box_shadow.dart [hunk 1]
```diff
@@ -20,65 +23,36 @@ import 'debug.dart';
...
   ///
   /// By default, the shadow is solid black with zero [offset], [blurRadius],
   /// and [spreadRadius].
   const BoxShadow({
-    this.color = const Color(0xFF000000),
-    this.offset = Offset.zero,
-    this.blurRadius = 0.0,
+    color = const Color(0xFF000000),
+    offset = Offset.zero,
+    blurRadius = 0.0,
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/box_shadow.dart]: these need types

File: packages/flutter/lib/src/painting/shadow.dart [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+/// A shadow.
+///
+/// See also:
+///
+///  * [Canvas.drawShadow], which is a more efficient way to draw shadows.
+///  * [BoxShadow], an extension of Shadow that includes a spreadRadius and can cast
+///    non-rectangular shadows.
+@immutable
+class Shadow {
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/shadow.dart]: ah, i would have expected this to be the class in dart:ui

**Review by chinmaygarde — APPROVED** [review]

**Comment by GaryQian** [comment]
If you are referencing the class `Shadow`, the idea is that `Shadow` is a base class of all sorts of faked shadows, and is not particularly specific to text. Other classes can presumably make use of it if they wished.

If you mean the 'shadows' property of TextStyle, since the "list of shadows" concept can be applied on multiple widgets, in order to keep it consistent and not have a <widgetname>Shadows prefix to each instance of shadows, I have called it just "shadows". I think this more unified API is a bit more predictable to work with. The fact that it is a Text shadow is signified by it being a property of TextStyle.

**Pull request merged**